File: gsdishopping_app/lib/models/coupon.dart
/// Algo que o cliente resgata gastando pontos, vindo do GSDIShoppingApi
/// (`GET /api/catalog/coupons`) — sempre pertence a uma campanha. Antes da
/// Fase 2 isto vinha do Squidex (`fromSquidex`); o Squidex saiu do projeto
/// (ver ARQUITETURA_TECNICA.md, seção 8) e o catálogo agora mora inteiro
/// no GSDIShoppingApi.
class Coupon {
  const Coupon({
    required this.id,
    required this.title,
    required this.description,
    required this.pointsCost,
    required this.campaignId,
    required this.campaignTitle,
    this.imageUrl,
    this.expiresAt,
  });

  final int id;
  final String title;
  final String description;
  final int pointsCost;
  final int campaignId;
  final String campaignTitle;
  final String? imageUrl;
  final DateTime? expiresAt;

  factory Coupon.fromJson(Map<String, dynamic> json) => Coupon(
        id: json['id'] as int,
        title: json['title'] as String,
        description: json['description'] as String,
        pointsCost: json['pointsCost'] as int,
        campaignId: json['campaignId'] as int,
        campaignTitle: json['campaignTitle'] as String,
        imageUrl: json['imageUrl'] as String?,
        expiresAt: json['expiresAt'] != null ? DateTime.tryParse(json['expiresAt'] as String) : null,
      );
}

File: GSDIShoppingApi/Models/Coupon.cs
namespace GSDIShoppingApi.Models;

/// <summary>
/// Algo que o cliente resgata/usa gastando pontos — sempre pertence a uma
/// <see cref="Campaign"/>. Isto é o que distingue Coupon de Promotion (ver
/// <see cref="Promotion"/>): cupom tem custo em pontos e um fluxo de
/// resgate (o resgate em si ainda não foi implementado — Fase 2 cobre só a
/// leitura do catálogo); promoção é só um anúncio de desconto de uma loja,
/// sem resgate.
/// </summary>
public class Coupon
{
    public int Id { get; set; }
    public required string Title { get; set; }
    public required string Description { get; set; }
    public required int PointsCost { get; set; }

    public required int CampaignId { get; set; }
    public Campaign? Campaign { get; set; }

    public string? ImageUrl { get; set; }
    public DateTime? ExpiresAt { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}
